cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll

```diff
@@ -0,0 +1,483 @@
+/**
+ * Provides sign analysis to determine whether expression are always positive
+ * or negative.
+ *
+ * The analysis is implemented as an abstract interpretation over the
+ * three-valued domain `{negative, zero, positive}`.
+ */
+import cpp
+private import semmle.code.cpp.ir.IR
+private import semmle.code.cpp.controlflow.IRGuards
+private import semmle.code.cpp.ir.ValueNumbering
+private import SignAnalysisCached
+
+private newtype TSign = TNeg() or TZero() or TPos()
+private class Sign extends TSign {
+  string toString() {
+    result = "-" and this = TNeg() or
+    result = "0" and this = TZero() or
+    result = "+" and this = TPos()
+  }
+  Sign inc() {
+    this = TNeg() and result = TNeg() or
+    this = TNeg() and result = TZero() or
+    this = TZero() and result = TPos() or
+    this = TPos() and result = TPos()
```

Comment:
This code assumes that overflow can't occur, but we often get FP reports for the existing range analysis library because overflow is relied on in practice. See, for example, https://discuss.lgtm.com/t/false-positive-c-comparison-result-is-always-the-same/1428.

I expect that we'll need to model overflow for unsigned integers, but we can continue to ignore it for signed integers because it's undefined there. For a case like `inc`, this means adding that `this = TPos() and result = TZero()` in the unsigned case.

Reply:
For the corresponding query in java, FPs for overflow checks are instead excluded in the query.  This works because overflow checks are pretty idiomatic and easy to recognise, and thus allows the rangeanalysis to stay (mostly) oblivious to such overflows.

Reply:
Okay, let's go with the Java approach for now and evaluate whether to change it later, when we have queries and results to look at.